compiler/src/dotty/tools/dotc/transform/CompleteJavaEnums.scala

```diff
@@ -145,17 +144,20 @@ class CompleteJavaEnums extends MiniPhase with InfoTransformer { thisPhase =>
    *          "same as before"
    *       }
    */
-  override def transformTemplate(templ: Template)(using Context): Template = {
+  override def transformTemplate(templ: Template)(using Context): Tree = {
     val cls = templ.symbol.owner
     if cls.derivesFromJavaEnum then
-      registerEnumClass(cls)
+      registerEnumClass(cls) // invariant: class is visited before cases: see tests/pos/enum-companion-first.scala
       val (params, rest) = decomposeTemplateBody(templ.body)
       val addedDefs = addedParams(cls, isLocal=true, ParamAccessor)
       val addedSyms = addedDefs.map(_.symbol.entered)
       val addedForwarders = addedEnumForwarders(cls)
       cpy.Template(templ)(
         parents = addEnumConstrArgs(defn.JavaEnumClass, templ.parents, addedSyms.map(ref)),
         body = params ++ addedDefs ++ addedForwarders ++ rest)
+    else if cls.linkedClass.derivesFromJavaEnum then
+      enumCaseOrdinals.clear() // remove simple cases // invariant: companion is visited after cases
```

Comment:
To make sure the map is always cleared, you could check that it is in empty in `checkPostCondition`

Reply:
Thanks for the tip, will do